ibm/resource_ibm_compute_vm_instance.go

```diff
@@ -495,18 +514,41 @@ func getVirtualGuestTemplateFromResourceData(d *schema.ResourceData, meta interf
 			return opts, fmt.Errorf("Error creating virtual guest: %s", err)
 		}
 		primaryNetworkComponent.NetworkVlan.PrimarySubnetId = &primarySubnetID
+		usePrimaryNetworkComponent = true
+	}
+
+	// Apply security groups if provided
+	publicSecurityGroups := d.Get("public_security_groups")
+	if publicSecurityGroups != nil {
+		publicSecurityGroupsList := publicSecurityGroups.(*schema.Set).List()
+		publicSGBindings := make([]datatypes.Virtual_Network_SecurityGroup_NetworkComponentBinding,
+			len(publicSecurityGroupsList))
+		for i, v := range publicSecurityGroupsList {
+			sgid, err := findSecurityGroupByName(v.(string), meta)
+			if err != nil {
+				return opts, fmt.Errorf("Error finding security group: %s", err)
+			}
+			publicSGBindings[i] = datatypes.Virtual_Network_SecurityGroup_NetworkComponentBinding{
+				SecurityGroupId: sl.Int(sgid),
+			}
+		}
```

Comment:
Since both `public_security_groups` and` private_security_groups` field have a common way of composing the 
SecurityGroupBindings , could you pull the logic in a function and use it across.

Typically such functions are expanders and you can call them 
expandSecurityGroupBindings or something like that. This function could return 
`[]datatypes.Virtual_Network_SecurityGroup_NetworkComponentBinding`
 

Note: Not needed as group names should be dropped and ids be used instead.